dulwich/porcelain.py

```diff
@@ -170,17 +170,21 @@ def open_repo_closing(path_or_repo):
     return closing(Repo(path_or_repo))
 
 
-def path_to_tree_path(repopath, path):
+def path_to_tree_path(repo_path, tree_path):
     """Convert a path to a path usable in e.g. an index.
 
-    :param repo: Repository
-    :param path: A path
-    :return: A path formatted for use in e.g. an index
+    :param repo_path: Absolute path to the repository
+    :param tree_path: Possibly non relative path to a file in the repository
+    :return: A path formatted for use in e.g. an index (bytes)
     """
-    os.path.relpath(path, repopath)
+    if tree_path.startswith(repo_path):
+        tree_path = os.path.relpath(tree_path, repo_path)
+    if isinstance(tree_path, str):
```

Comment:
This is invalid on Python2.

Reply:
which part ? `startswith` is defined, as is `isinstance`, no ?

Reply:
the isinstance bit; if something is a bytestring on python2 and it has non-ascii characters, then encoding it will lead to mayhem since it tries to convert from ascii:

>>> "ĳ".encode('utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc4 in position 0: ordinal not in range(128)

Reply:
ah right, then we can just do the opposite: `not isinstance(f(path), bytes)`. just checked it seems to work in python2